src/KubernetesClient/Informers/SharedInformer.cs

```diff
@@ -0,0 +1,184 @@
+    public class SharedInformer<TKey, TResource> : IInformer<TResource>
+    {
+        private readonly ICache<TKey, TResource> _cache;
+        private readonly Func<TResource, TKey> _keySelector;
+        private readonly object _lock = new object();
+        private readonly ILogger _logger;
+        private readonly CountdownEvent _waitingSubscribers = new CountdownEvent(0);
+        private TaskCompletionSource<bool> _cacheSynchronized = new TaskCompletionSource<bool>();
+        private readonly IConnectableObservable<CacheSynchronized<ResourceEvent<TResource>>> _masterObservable;
+        private readonly IScheduler _masterScheduler;
+        private IDisposable _masterSubscription;
+        private int _subscribers;
+
+        public SharedInformer(IInformer<TResource> masterInformer, ILogger logger, Func<TResource, TKey> keySelector)
+            : this(masterInformer, logger, keySelector, new SimpleCache<TKey, TResource>())
+        {
+        }
+
+        public SharedInformer(IInformer<TResource> masterInformer, ILogger logger, Func<TResource, TKey> keySelector, ICache<TKey, TResource> cache, IScheduler scheduler = null)
+        {
+            _cache = cache;
+            _masterScheduler = scheduler ?? new EventLoopScheduler();
+            _logger = logger;
+            _keySelector = keySelector;
+            _masterObservable = masterInformer
+                .GetResource(ResourceStreamType.ListWatch)
+                .ObserveOn(_masterScheduler)
+                .Do(x => _logger.LogTrace($"Received message from upstream {x}"))
+                .SynchronizeCache(_cache, _keySelector)
+                .Do(msg =>
+                {
+                    // cache is synchronized as soon as we get at least one message past this point
+                    _logger.LogTrace($"Cache v{cache.Version} synchronized: {msg} ");
+                    _cacheSynchronized.TrySetResult(true);
+                    _logger.LogTrace("_cacheSynchronized.TrySetResult(true)");
+                })
+                .Do(_ => YieldToWaitingSubscribers())
+                .ObserveOn(Scheduler.Immediate) // immediate ensures that all caches operations are done atomically
+                .ObserveOn(_masterScheduler)
+                .Catch<CacheSynchronized<ResourceEvent<TResource>>, Exception>(e =>
+                {
+                    _cacheSynchronized.TrySetException(e);
+                    // _cacheSynchronized.OnError(e);
+                    return Observable.Throw<CacheSynchronized<ResourceEvent<TResource>>>(e);
+                })
+                .Finally(() => _cacheSynchronized.TrySetResult(false))
+                // .SubscribeOn(_masterScheduler)
+                .Publish();
+        }
+
+
+        public IObservable<ResourceEvent<TResource>> GetResource(ResourceStreamType type)
+        {
+            var childScheduler = new EventLoopScheduler(); // dedicated thread for the child on which all messages are syncrhonized
```

Comment:
The scheduler must be disposed; otherwise, its thread [waits](https://github.com/dotnet/reactive/blob/e00d53bd022dd2a787d983e0fbbcd04016978bc2/Rx.NET/Source/src/System.Reactive/Concurrency/EventLoopScheduler.cs#L298) indefinitely for an event that will never be signalled.

Reply:
I'm looking for a little help on how to do this properly. Adding it as part of observable `Dispose` lifeline terminates it too early and breaks the whole thing. A bit of a dirty hack would be to put it on something like Task.Delay(5000) to allow the stream to finalize and close before the actual thread is killed off. Let me know if you have a better idea.

Reply:
Might `Observable.Using` help?

Reply:
That did it. I was not aware of that operator.